internal/uvm/cpugroups.go

```diff
@@ -0,0 +1,158 @@
+}
+
+// setCPUGroup sets the VM's cpugroup
+func (uvm *UtilityVM) setCPUGroup(ctx context.Context, id string) error {
+	req := &hcsschema.ModifySettingRequest{
+		ResourcePath: cpuGroupResourcePath,
+		Settings: &hcsschema.CpuGroup{
+			Id: id,
+		},
+	}
+	if err := uvm.modify(ctx, req); err != nil {
+		return err
+	}
+	uvm.cpuGroupID = id
+	return nil
+}
+
+// unsetCPUGroup sets the VM's cpugroup to the null group ID
+// set groupID to 00000000-0000-0000-0000-000000000000 to remove the VM from a cpugroup
+//
+// Since a VM must be moved to the null group before potentially being added to a different
+// cpugroup, that means there may be a segment of time that the VM's cpu usage runs unrestricted.
+func (uvm *UtilityVM) unsetCPUGroup(ctx context.Context) error {
+	return uvm.setCPUGroup(ctx, cpuGroupNullID)
+}
+
+// DeleteCPUGroup deletes the cpugroup from the host
+func DeleteCPUGroup(ctx context.Context, id string) error {
+	operation := hcsschema.DeleteGroup
+	details := hcsschema.DeleteGroupOperation{
+		GroupId: id,
+	}
+
+	return modifyCPUGroupRequest(ctx, operation, details)
+}
+
+// modifyCPUGroupRequest is a helper function for making modify calls to a cpugroup
+func modifyCPUGroupRequest(ctx context.Context, operation hcsschema.CPUGroupOperation, details interface{}) error {
+	req := hcsschema.ModificationRequest{
+		PropertyType: hcsschema.PTCPUGroup,
+		Settings: &hcsschema.HostProcessorModificationRequest{
+			Operation:        operation,
+			OperationDetails: details,
+		},
+	}
+
+	return hcs.ModifyServiceSettings(ctx, req)
+}
+
+// CreateNewCPUGroupWithID creates a new cpugroup on the host with a prespecified id
+func CreateNewCPUGroupWithID(ctx context.Context, id string, logicalProcessors []uint32) error {
+	operation := hcsschema.CreateGroup
+	details := &hcsschema.CreateGroupOperation{
+		GroupId:               strings.ToLower(id),
+		LogicalProcessors:     logicalProcessors,
+		LogicalProcessorCount: uint32(len(logicalProcessors)),
+	}
+	if err := modifyCPUGroupRequest(ctx, operation, details); err != nil {
+		return fmt.Errorf("failed to make cpugroups CreateGroup request with details %v with: %s", details, err)
+	}
+	return nil
+}
+
+// getHostCPUGroups queries the host for cpugroups and their properties.
+func getHostCPUGroups(ctx context.Context) (*hcsschema.CpuGroupConfigurations, error) {
+	query := hcsschema.PropertyQuery{
+		PropertyTypes: []hcsschema.PropertyType{hcsschema.PTCPUGroup},
+	}
+
+	cpuGroupsPresent, err := hcs.GetServiceProperties(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+
+	groupConfigs := &hcsschema.CpuGroupConfigurations{}
+	if err := json.Unmarshal(cpuGroupsPresent.Properties[0], groupConfigs); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal host cpugroups: %v", err)
+	}
+
+	return groupConfigs, nil
+}
+
+// getCPUGroupConfig finds the cpugroup config information for group with `id`
+func getCPUGroupConfig(ctx context.Context, id string) (*hcsschema.CpuGroupConfig, error) {
```

Comment:
It seems like this is only called in `cpuGroupExists` to check if the group exists and the config isn't actually used thereafter. Could we just consolidate these two?

Reply:
fixed